conformance/lang/expressions/type_cast_expression.balt

```diff
@@ -0,0 +1,1698 @@
+             applicable by casting to an object type.
+Labels: object-constructor-expr, object-type, type-cast-expr, var
+
+function errorFunction() {
+    var a = object {
+        int a = 5;
+    };
+
+    object {int b;} _ = <object {int b;}> a; // @error incompatible field names
+}
+
+Test-Case: error
+Description: Test error when value is not belonging to the type described by the type desc and numeric convert is not 
+             applicable by casting to a stream type.
+Labels: additive-expr, implicit-new-expr, isolated-method, mapping-constructor-expr, module-class-defn, optional-type, 
+        record-type, stream, stream-type-parameters, string, type-cast-expr
+
+class NumberGenerator {
+    int number = 0;
+    public isolated function next() returns record {|int value;|}? {
+        self.number = self.number + 1;
+        return {value: self.number};
+    }
+}
+
+function errorFunction() {
+    NumberGenerator numberGen = new ();
+    stream<int> streamName = new (numberGen);
+    stream<string> _ = <stream<string>>streamName; // @error 'stream<int>' cannot be cast to 'stream<string>'
+}
+
+Test-Case: error
+Description: Test error when attempting to use a type cast to cast away an error.
+Labels: error, module-type-defn, string, type-cast-expr, union-type
+
+type MyError error;
+
+type MyUnion int|MyError;
+
+function errorFunction() {
+    int|error a = 1;
+    int b = <int>a; // @error attempt to cast away an error
+
+    string c = <string>getString(); // @error attempt to cast away an error
+
+    MyUnion d = 3;
+    int e = <int>d; // @error attempt to cast away an error
+}
+
+function getString() returns error|string {
+    return "car";
+}
+
+Test-Case: output
+Description: Test if static type of expression contains error, then the intersection of T and error must 
+             also be non-empty.
+Labels: error, error-constructor-expr, error-type-parameter, field-access-expr, method-call-expr, module-type-defn, 
+        record-type, string, type-cast-expr, union-type
+
+type MyError error<record {|int severity;|}>;
+
+type MyUnion any|MyError;
+
+function init() {
+    MyUnion a = 3;
+    int|error b = <int|error>a;
+    io:println(b); // @output 3
+
+    string|error c = error("msg", severity = 1);
+    MyError e = <MyError> c;
+    io:println(e.detail().severity); // @output 1
+}
+
+Test-Case: output
+Description: Test the contextually expected type for expression is the intersection of the contextually expected type 
+             of the type-cast-expr and the type described by the type-descriptor
+Labels: any, array-type, byte, decimal, DecimalFloatingPointNumber, float, mapping-constructor-expr, 
+        record-rest-descriptor, record-type, record-type-readonly-field, string, tuple-type, type-cast-expr, union-type
+
+function init() {
+    record {|int a;string...; |} a = <record {|int a;|}>{a: 5};
+    io:println(typeof a); // @output typedesc record {| int a; |}
```

Comment:
`typeof` checks the runtime type?

I guess we can also add a negative test like
```ballerina
function init() {
    string[] a = [];
    int[] c = <readonly>a; // incompatible types: 'string[]' cannot be cast to '(int[] & readonly)'
}
```

Reply:
This test checks the contextually expected type for the expression in the type cast. To test that the expression should be a constructor right?. The example provided by u is not related to that